c2c: skip and deflake a few unit tests

msbutler opened the pull request:
c2c: skip a few unit tests
Informs #102523
Informs #102286
Informs #86206

Release note: none

c2c: close pgconn with correct context in test infra
When a user opens a pgx connection, they pass it a context that must be used to
close the connection, else pgx's contextWatcher will run indefinetly on a
seperate goroutine.  Previously, the `startReplication()` test helper was
initializing its own context to open pgx connection, and did not pass this
context back to the caller, and consequently, the user would close the
connection with the incorrect context, leading to a leaky goroutine.

This patch allows the user to pass a context in the startReplication(),
preventing the user from closing the connection with the wrong goroutine.

Fixes https://github.com/cockroachdb/cockroach/issues/102523

Release note: None

msbutler committed:
c2c: skip a few unit tests

Informs #102523
Informs #102286
Informs #86206

Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/102593)
<!-- Reviewable:end -->

adityamaru commented on pkg/util/leaktest/leaktest.go:
```diff
@@ -58,8 +58,9 @@ func interestingGoroutines() map[int64]string {
-			// Ignore pgconn which creates a goroutine to do an async cleanup.
+			// Ignore pgconn which creates two goroutines to do async cleanup.
 			strings.Contains(stack, "github.com/jackc/pgconn.(*PgConn).asyncClose.func1") ||
+			strings.Contains(stack, "github.com/jackc/pgconn/internal/ctxwatch.(*ContextWatcher).Watch.func1()") ||
```
what is this about? just want to check its intentional

dt replied:
we should be very sure this isn't a missing Close() before we add it to the list or we could be hiding a real drain / shutdown bug

msbutler replied:
good call. i'll investigate further and skip the flaky test for now.

msbutler committed:
c2c: close pgconn with correct context in test infra

When a user opens a pgx connection, they pass it a context that must be used to
close the connection, else pgx's contextWatcher will run indefinetly on a
seperate goroutine.  Previously, the `startReplication()` test helper was
initializing its own context to open pgx connection, and did not pass this
context back to the caller, and consequently, the user would close the
connection with the incorrect context, leading to a leaky goroutine.

This patch allows the user to pass a context in the startReplication(),
preventing the user from closing the connection with the wrong goroutine.

Fixes #102523

Release note: None

msbutler commented:
@dt I think I found the test infra bug. I think we were mishandling contexts. I wonder if I should revert https://github.com/cockroachdb/cockroach/pull/102258 as well?

msbutler committed:
Revert "leaktest: ignore pgconn async close worker"

This reverts commit 8c2ae78f85ed24dd5a943694e22731774c1aa18a.

msbutler commented:
bors r=benbardin

msbutler commented:
bors retry

msbutler commented:
blathers backport 23.1
